File: src/C#/StructuralPatterns/Adapter/ThirdPartyBillingSystem/ThirdPartyBillingSystemV2.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ThirdPartyBillingSystemV2
{
    public class Employee
    {
        public int ID { get; set; }

        public string Name { get; set; }

        public string Designation { get; set; }

        public decimal Salary { get; set; }
    }

    public interface IEmployeesTarget
    {
        void ProcessCompanySalary(List<Employee> employees);
    }

    public class EmployeesAdapter : IEmployeesTarget
    {
        private ThirdPartyBillingSystem thirdPartyBillingSystem = new ThirdPartyBillingSystem();

        public void ProcessCompanySalary(List<Employee> employees)
        {
            string[][] adaptedEmployeesJaggedArray = employees
                .Select(e => new string[] { e.ID.ToString(), e.Name, e.Designation, e.Salary.ToString() })
                .ToArray();

            Console.WriteLine("Adapter converted array of employees to a list of employees");
            Console.WriteLine("Then delegate to the ThirdPartyBillingSystem for processing the employees salary\n");

            thirdPartyBillingSystem.ProcessSalary(adaptedEmployeesJaggedArray);
        }
    }

    public class ThirdPartyBillingSystem
    {
        public void ProcessSalary(string[][] employeesJaggedArray)
        {
            foreach (var employeeArray in employeesJaggedArray)
            {
                Console.WriteLine($"Salary: {employeeArray[3]}, " +
                    $"Credited to: {employeeArray[1]} with designation: {employeeArray[2]}");
            }
        }
    }

    public class Engine
    {
        // Enter data from EmployeesV2.txt

        public static void Run()
        {
            List<Employee> employeesToProcessCompanySalary = new List<Employee>();

            string employeeCommand = Console.ReadLine();

            while (employeeCommand != "END")
            {
                string[] employeeCommands = employeeCommand.Split().ToArray();

                var employeeToProcessCompanySalary = new Employee
                {
                    ID = int.Parse(employeeCommands[0]),
                    Name = employeeCommands[1],
                    Designation = employeeCommands[2],
                    Salary = decimal.Parse(employeeCommands[3])
                };

                employeesToProcessCompanySalary.Add(employeeToProcessCompanySalary);

                employeeCommand = Console.ReadLine();
            }

            IEmployeesTarget employeesTarget = new EmployeesAdapter();
            employeesTarget.ProcessCompanySalary(employeesToProcessCompanySalary);
        }
    }
}

File: src/Java/StructuralPatterns/Adapter/ThirdPartyBillingSystem/src/ThirdPartyBillingSystemV2.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class EmployeeV2 {
    private int ID;
    private String Name;
    private String Designation;
    private double Salary;

    public int getID() {
        return ID;
    }

    public void setID(int ID) {
        this.ID = ID;
    }

    public String getName() {
        return Name;
    }

    public void setName(String Name) {
        this.Name = Name;
    }

    public String getDesignation() {
        return Designation;
    }

    public void setDesignation(String Designation) {
        this.Designation = Designation;
    }

    public double getSalary() {
        return Salary;
    }

    public void setSalary(double Salary) {
        this.Salary = Salary;
    }

    public EmployeeV2(int ID, String Name, String Designation, double Salary) {
        this.ID = ID;
        this.Name = Name;
        this.Designation = Designation;
        this.Salary = Salary;
    }
}

interface IEmployeesTargetV2 {
    void ProcessCompanySalary(List<EmployeeV2> employees);
}

class EmployeesAdapterV2 implements IEmployeesTargetV2 {
    private ThirdPartyBillingSystemV2 thirdPartyBillingSystem = new ThirdPartyBillingSystemV2();

    @Override
    public void ProcessCompanySalary(List<EmployeeV2> employees) {
        String[][] adaptedEmployeesJaggedArray = employees
                .stream()
                .map(e -> new String[]{String.valueOf(e.getID()), e.getName(), e.getDesignation(), String.valueOf(e.getSalary())})
                .toArray(String[][]::new);

        System.out.println("Adapter converted list of employees to a 2D array");
        System.out.println("Then delegate to the ThirdPartyBillingSystem for processing the employees salary\n");

        thirdPartyBillingSystem.ProcessSalary(adaptedEmployeesJaggedArray);
    }
}

class ThirdPartyBillingSystemV2 {
    public void ProcessSalary(String[][] employeesJaggedArray) {
        for (String[] employeeArray : employeesJaggedArray) {
            System.out.println("Salary: " + employeeArray[3] + ", " +
                    "Credited to: " + employeeArray[1] + " with designation: " + employeeArray[2]);
        }
    }
}

class EngineV2 {
    public static void run() {
        List<EmployeeV2> employeesToProcessCompanySalary = new ArrayList<>();

        Scanner scanner = new Scanner(System.in);
        String employeeCommand = scanner.nextLine();

        while (!employeeCommand.equals("END")) {
            String[] employeeCommands = employeeCommand.split(" ");

            EmployeeV2 employeeToProcessCompanySalary = new EmployeeV2(
                    Integer.parseInt(employeeCommands[0]),
                    employeeCommands[1],
                    employeeCommands[2],
                    Double.parseDouble(employeeCommands[3])
            );

            employeesToProcessCompanySalary.add(employeeToProcessCompanySalary);

            employeeCommand = scanner.nextLine();
        }

        IEmployeesTargetV2 employeesTarget = new EmployeesAdapterV2();
        employeesTarget.ProcessCompanySalary(employeesToProcessCompanySalary);
    }
}